jmh/jmh.bzl

```diff
@@ -0,0 +1,149 @@
+  )
+  native.bind(
+      name = "io_bazel_rules_scala/dependency/jmh/net_sf_jopt_simple_jopt_simple",
+      actual = '@io_bazel_rules_scala_net_sf_jopt_simple_jopt_simple//jar',
+  )
+  native.maven_jar(
+      name = "io_bazel_rules_scala_org_apache_commons_commons_math3",
+      artifact = "org.apache.commons:commons-math3:3.6.1",
+      sha1 = "e4ba98f1d4b3c80ec46392f25e094a6a2e58fcbf",
+  )
+  native.bind(
+      name = "io_bazel_rules_scala/dependency/jmh/org_apache_commons_commons_math3",
+      actual = '@io_bazel_rules_scala_org_apache_commons_commons_math3//jar',
+  )
+
+
+def scala_benchmark_jmh(**kw):
+  name = kw["name"]
+  deps = kw.get("deps", [])
+  srcs = kw["srcs"]
+  lib = "%s_generator" % name
+  scalacopts = kw.get("scalacopts", [])
+  main_class = kw.get("main_class", "org.openjdk.jmh.Main")
+
+  scala_library(
+      name = lib,
+      srcs = srcs,
+      deps = deps + [
+          "//external:io_bazel_rules_scala/dependency/jmh/jmh_core",
+      ],
+      scalacopts = scalacopts,
+      visibility = ["//visibility:public"],
+  )
+  jmh_benchmark_generator_tool = "__jmh_benchmark_generator"
+  scala_binary(
+    name = jmh_benchmark_generator_tool,
+    main_class = "io.bazel.rules_scala.jmh_support.BenchmarkGenerator",
+    deps = [
+      "//src/scala/io/bazel/rules_scala/jmh_support:benchmark_generator_lib"
+    ],
+  )
+
+  codegen = name + "_codegen"
+  src_jar = "%s_jmh.srcjar" % name
+  benchmark_list = "resources/META-INF/BenchmarkList"
+  compiler_hints = "resources/META-INF/CompilerHints"
+
+  binary_deploy_jar = "//src/java/io/bazel/rulesscala/jar:binary_deploy.jar"
+
+  java_tool = "@local_jdk//:java"
+  jar_tool = "@local_jdk//:jar"
+  jdk_tool = "//tools/defaults:jdk"
+
+  native.genrule(
+      name = codegen,
+      srcs = [lib, binary_deploy_jar],
+      outs = [src_jar, benchmark_list, compiler_hints],
+      tools = [
+          jmh_benchmark_generator_tool,
+          java_tool,
+          jar_tool,
+          jdk_tool,
+      ],
+      cmd = """
+BINARY_DEPLOY_JAR=`pwd`/$(location {binary_deploy_jar})
+JAVA=`pwd`/$(location {java_tool})
+JAR=`pwd`/$(location {jar_tool})
+
+pushd `dirname $(location {lib})`
+$$JAR -xf `basename $(location {lib})`
+popd
+
+./$(location {jmh_benchmark_generator_tool}) $(location {lib}) $(@D)
+pushd $(@D)/sources
+$$JAVA -jar $$BINARY_DEPLOY_JAR ../{name}_jmh.srcjar ./
+popd
+""".format(
+      jmh_benchmark_generator_tool=jmh_benchmark_generator_tool,
+      binary_deploy_jar=binary_deploy_jar,
+      java_tool=java_tool,
+      jar_tool=jar_tool,
+      lib=lib,
+      name=name,
+    ),
+  )
+
+  compiled_lib = name + "_compiled_benchmark_lib"
+  scala_library(
+      name = compiled_lib,
+      srcs = [src_jar],
+      deps = deps + [
+          "//external:io_bazel_rules_scala/dependency/jmh/jmh_core",
+          lib,
+          codegen,
+      ],
+      resources = [benchmark_list, compiler_hints],
+  )
+  scala_binary(
```

Comment:
After discussing, it sounds like it makes the most sense to leave as-is, with the build product being a runnable benchmark.